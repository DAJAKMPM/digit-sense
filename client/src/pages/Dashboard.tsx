import React, { useCallback, useEffect, useState } from "react";
import { useNavigate } from "react-router-dom";
import {
  Box,
  Button,
  Paper,
  Stack,
  Table,
  TableBody,
  TableCell,
  TableContainer,
  TableFooter,
  TablePagination,
  TableRow,
} from "@mui/material";
import EditIcon from "@mui/icons-material/Edit";
import DeleteIcon from "@mui/icons-material/Delete";
import AddIcon from "@mui/icons-material/Add";
import { Layout } from "../components/Layout";
import { Pagination } from "../components/Pagination";
import { User } from "../types";
import { deleteUser, getAllUsers } from "../services";

const DashboardScreen = () => {
  const [users, setUsers] = useState<User[]>([]);
  const [activePage, setActivePage] = useState<number>(0);
  const [totalCount, setTotalCount] = useState<number>(0);

  const navigate = useNavigate();

  const handleFetchUsers = useCallback(async () => {
    try {
      const currentPage = activePage === 0 ? 1 : activePage + 1;
      const { data, totalCount, success } = await getAllUsers(currentPage);

      if (success) {
        setUsers(data);
        setTotalCount(totalCount);
      }
    } catch (error) {
      console.log(error);
    }
  }, [activePage]);

  useEffect(() => {
    handleFetchUsers();
  }, [handleFetchUsers]);

  const handleEditUser = (id?: string) => {
    if (!id) return;

    navigate(`/user/${id}`);
  };
  const handleCreateUser = () => navigate("/create-user");

  const handleDeleteUser = async (id?: string) => {
    if (!id) return;

    try {
      const { success } = await deleteUser(id);

      if (success) handleFetchUsers();
    } catch (error) {
      console.log(error);
    }
  };

  const handleChangePage = (
    _event: React.MouseEvent<HTMLButtonElement> | null,
    newPage: number
  ) => {
    setActivePage(newPage);
  };

  return (
    <Layout pageTitle="Dashboard">
      <Box height={400} width="100%" mt={4}>
        <Box width="100%" sx={{ display: "flex", justifyContent: "flex-end" }}>
          <Button
            variant="contained"
            startIcon={<AddIcon />}
            onClick={handleCreateUser}
            sx={{ mt: 3, mb: 2 }}
          >
            Add new user
          </Button>
        </Box>

        <TableContainer component={Paper}>
          <Table sx={{ minWidth: 500 }} aria-label="custom pagination table">
            <TableBody>
              {users.map((row) => (
                <TableRow key={row.name}>
                  <TableCell align="left">{row._id}</TableCell>
                  <TableCell align="left">{row.email}</TableCell>
                  <TableCell align="left">{row.name}</TableCell>
                  <TableCell align="left">{row.address}</TableCell>
                  <TableCell align="left">{row.age}</TableCell>
                  <TableCell>
                    <Stack direction="row" spacing={2}>
                      <Button
                        variant="contained"
                        startIcon={<EditIcon />}
                        onClick={() => handleEditUser(row._id)}
                      >
                        Edit
                      </Button>
                      <Button
                        variant="outlined"
                        color="error"
                        startIcon={<DeleteIcon />}
                        onClick={() => handleDeleteUser(row._id)}
                      >
                        Delete
                      </Button>
                    </Stack>
                  </TableCell>
                </TableRow>
              ))}
            </TableBody>
            <TableFooter>
              <TableRow>
                <TablePagination
                  rowsPerPageOptions={[2]}
                  colSpan={3}
                  count={totalCount}
                  rowsPerPage={5}
                  page={activePage}
                  onPageChange={handleChangePage}
                  ActionsComponent={Pagination}
                />
              </TableRow>
            </TableFooter>
          </Table>
        </TableContainer>
      </Box>
    </Layout>
  );
};

export { DashboardScreen };
