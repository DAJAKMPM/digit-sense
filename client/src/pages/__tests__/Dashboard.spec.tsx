import { render } from "@testing-library/react";
import { BrowserRouter as Router } from "react-router-dom";
import { DashboardScreen } from "../Dashboard";

test("should match the DashboardScreen snapshot properly", () => {
  const { container } = render(
    <Router>
      <DashboardScreen />
    </Router>
  );

  expect(container).toMatchSnapshot();
});
