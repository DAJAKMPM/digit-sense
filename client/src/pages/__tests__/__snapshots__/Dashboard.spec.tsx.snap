// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match the DashboardScreen snapshot properly 1`] = `
<div>
  <main>
    <header
      class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation0 MuiAppBar-root MuiAppBar-colorDefault MuiAppBar-positionAbsolute css-3e2rzi-MuiPaper-root-MuiAppBar-root"
    >
      <div
        class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-hyum1k-MuiToolbar-root"
      >
        <h6
          class="MuiTypography-root MuiTypography-h6 MuiTypography-noWrap css-uh02xs-MuiTypography-root"
        >
          Dashboard
        </h6>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textInherit MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorInherit css-1y942vo-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          Profile
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textInherit MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorInherit css-1y942vo-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          Logout
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </header>
    <main
      class="MuiContainer-root MuiContainer-maxWidthLg css-2kezlg-MuiContainer-root"
    >
      <div
        class="MuiBox-root css-u4ugex"
      >
        <div
          class="MuiBox-root css-2ertzu"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-1n7zltw-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiButton-startIcon MuiButton-iconSizeMedium css-1d6wzja-MuiButton-startIcon"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="AddIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
                />
              </svg>
            </span>
            Add new user
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
        <div
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiTableContainer-root css-11xur9t-MuiPaper-root-MuiTableContainer-root"
        >
          <table
            aria-label="custom pagination table"
            class="MuiTable-root css-130xbax-MuiTable-root"
          >
            <thead
              class="MuiTableHead-root css-15wwp11-MuiTableHead-root"
            >
              <tr
                class="MuiTableRow-root MuiTableRow-head css-1q1u3t4-MuiTableRow-root"
              >
                <th
                  class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignLeft MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
                  scope="col"
                >
                  User ID
                </th>
                <th
                  class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignLeft MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
                  scope="col"
                >
                  Email
                </th>
                <th
                  class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignLeft MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
                  scope="col"
                >
                  Full Name
                </th>
                <th
                  class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignLeft MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
                  scope="col"
                >
                  Address
                </th>
                <th
                  class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignLeft MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
                  scope="col"
                >
                  Age
                </th>
                <th
                  class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1ndpvdd-MuiTableCell-root"
                  scope="col"
                >
                  Actions
                </th>
              </tr>
            </thead>
            <tbody
              class="MuiTableBody-root css-apqrd9-MuiTableBody-root"
            />
            <tfoot
              class="MuiTableFooter-root css-9cgc89-MuiTableFooter-root"
            >
              <tr
                class="MuiTableRow-root MuiTableRow-footer css-1q1u3t4-MuiTableRow-root"
              >
                <td
                  class="MuiTableCell-root MuiTableCell-footer MuiTableCell-sizeMedium MuiTablePagination-root css-fikjyc-MuiTableCell-root-MuiTablePagination-root"
                  colspan="3"
                >
                  <div
                    class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular MuiTablePagination-toolbar css-78c6dr-MuiToolbar-root-MuiTablePagination-toolbar"
                  >
                    <div
                      class="MuiTablePagination-spacer css-1psng7p-MuiTablePagination-spacer"
                    />
                    <p
                      class="MuiTablePagination-displayedRows css-levciy-MuiTablePagination-displayedRows"
                    >
                      0â€“0 of 0
                    </p>
                    <div
                      class="MuiBox-root css-1zye22"
                    >
                      <button
                        aria-label="previous page"
                        class="MuiButtonBase-root Mui-disabled MuiIconButton-root Mui-disabled MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
                        disabled=""
                        tabindex="-1"
                        type="button"
                      >
                        <svg
                          aria-hidden="true"
                          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                          data-testid="KeyboardArrowLeftIcon"
                          focusable="false"
                          viewBox="0 0 24 24"
                        >
                          <path
                            d="M15.41 16.59 10.83 12l4.58-4.59L14 6l-6 6 6 6 1.41-1.41z"
                          />
                        </svg>
                      </button>
                      <button
                        aria-label="next page"
                        class="MuiButtonBase-root Mui-disabled MuiIconButton-root Mui-disabled MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
                        disabled=""
                        tabindex="-1"
                        type="button"
                      >
                        <svg
                          aria-hidden="true"
                          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                          data-testid="KeyboardArrowRightIcon"
                          focusable="false"
                          viewBox="0 0 24 24"
                        >
                          <path
                            d="M8.59 16.59 13.17 12 8.59 7.41 10 6l6 6-6 6-1.41-1.41z"
                          />
                        </svg>
                      </button>
                    </div>
                  </div>
                </td>
              </tr>
            </tfoot>
          </table>
        </div>
      </div>
    </main>
  </main>
</div>
`;
